### ISSUE TITLE
Support type-only TypedDocumentNode declarations

### ISSUE THREAD

**Issue by chriskrycho** [dotansimha/graphql-code-generator#5043]
Currently, given a query like this—

```graphql
query findUser($userId: ID!) {
    user(id: $userId) {
        id
        username
        role
    }
}
```

—the generator creates this output (assuming `typescript-operations` and `near-operation-file` as well for convenience):

```ts
import * as Types from './types';

import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type FindUserQueryVariables = Types.Exact<{
  userId: Types.Scalars['ID'];
}>;


export type FindUserQuery = { __typename?: 'Query', user?: Types.Maybe<{ __typename?: 'User', id: string, username: string, role: Types.Role }> };


export const FindUserDocument: DocumentNode<FindUserQuery, FindUserQueryVariables> = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"findUser"},"variableDefinitions":[{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"userId"}},"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"ID"}}},"directives":[]}],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"user"},"arguments":[{"kind":"Argument","name":{"kind":"Name","value":"id"},"value":{"kind":"Variable","name":{"kind":"Name","value":"userId"}}}],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"},"arguments":[],"directives":[]},{"kind":"Field","name":{"kind":"Name","value":"username"},"arguments":[],"directives":[]},{"kind":"Field","name":{"kind":"Name","value":"role"},"arguments":[],"directives":[]}]}}]}}]};
```

For many purposes, however, there's zero value in the *actual* TS output, only a type definition, and we in fact want to generate only a `.d.ts` file, which should look something like this:

```ts
import * as Types from './types';

import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type FindUserQueryVariables = Types.Exact<{
  userId: Types.Scalars['ID'];
}>;


export type FindUserQuery = { __typename?: 'Query', user?: Types.Maybe<{ __typename?: 'User', id: string, username: string, role: Types.Role }> };


export const FindUserDocument: DocumentNode<FindUserQuery, FindUserQueryVariables>;
```

**Comment by bennypowers** [comment]
Just chiming in to express my interest in this feature. It would enable me to [add `TypedDocumentNode` support to apollo elements](https://github.com/apollo-elements/apollo-elements/pull/96)


```ts
// import { HelloQueryData, HelloQueryVariables } from '#schema'; // old style codegen
import { HelloQueryDocument } from '#schema'; // after this issue is merged

import HelloQuery from './Hello.query.graphql'; // actual `.graphql` file, transformed with a build step or dev-server plugin

// old style codegen, but still supported with some type inference
class OldStyleHelloQueryElement extends ApolloQuery<HelloQueryData, HelloQueryVariables> {
  query = HelloQuery;
}

// new style, accepting a TypedDocumentNode
class HelloQueryElement extends ApolloQuery<HelloQueryDocument> {
  query = HelloQuery;
}
```

OR

```ts
import { HelloQueryDocument } from '#schema'; // full codegen
// OR
import { HelloQueryDocument } from './Hello.query'; // file codegen

// also supported via `typeof`
class HelloQueryElement extends ApolloQuery<typeof HelloQueryDocument> {
  query = HelloQueryDocument;
}
```

**Comment by benquarmby** [comment]
TL:DR: Is it worth enhancing [the `addOperationExport` feature](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-operations#addoperationexport) instead?

Hello from five years in the future! I also need a declaration only version of the `typed-document-node` plugin. Back in 2020, a somewhat related feature was added to the `typescript-operations` plugin:
- #4020

Given this fabricated query:

```gql
# getUser.gql
query getUser {
    currentUser {id}
}
```

A declaration only, `near-operation-file` setup with `addOperationExport` might generate something like this. The export naming lines up with [how `graphql-tag/laoder` works](https://github.com/apollographql/graphql-tag?tab=readme-ov-file#webpack-loading-and-preprocessing) in webpack:

```ts
// getUser.d.gql.ts
import * as Types from "../schema.generated.js";

export type GetUserQueryVariables = Types.Exact<{[key: string]: never}>;

export type GetUserQuery = {
    __typename?: "Query";
    currentUser?: {
        __typename?: "User";
        id: string;
    } | null;
};

// This is the most important line for this discussion 👇 
export declare const getUser: import("graphql").DocumentNode;
```

This gets us pretty close, but you can tell straight away it's not very strongly typed. What we really want is an option to do this:

```diff
- export declare const getUser: import("graphql").DocumentNode;
+ export declare const getUser: import("@graphql-typed-document-node/core").TypedDocumentNode<GetUserQuery, GetUserQueryVariables>;
```

And as an added bonus, it should also support the first-document-is-the-default-export behavior of `graphql-tag/loader`:

```ts
export default getUser;
```

I have a strawman running locally that does the above, but it feels very hacky and not the graphql-codgen-way™. I would probably need some help to tighten it up, assuming there is interest in pursuing `addOperationExport` as an alternative?

**Comment by benquarmby** [comment]
Here is a strawman plugin. It's very basic without support for naming conventions, anonymous queries etc. But it shows the concept:

```js
// graphql-typed-exports-plugin.mjs
import {pascalCase} from "change-case-all";

export function plugin(schema, documents) {
    let firstExport;

    const outputs = documents.map(function ({document}, index) {
        const [rootDefinition] = document.definitions;
        const documentName = rootDefinition.name.value;
        const outputType = pascalCase(documentName) + pascalCase(rootDefinition.operation);
        const inputType = outputType + "Variables";

        if (index === 0) {
            firstExport = documentName;
        }

        return `export declare const ${documentName}: import("@graphql-typed-document-node/core").TypedDocumentNode<${outputType}, ${inputType}>;`;
    });

    if (firstExport) {
        outputs.push(`export default ${firstExport};`);
    }

    return outputs.join("\n");
}
```
